module/zfs/zfs_fm.c

```diff
@@ -931,9 +939,60 @@ zfs_post_autoreplace(spa_t *spa, vdev_t *vd)
  * open because the device was not found (fault.fs.zfs.device).
  */
 void
-zfs_post_state_change(spa_t *spa, vdev_t *vd)
+zfs_post_state_change(spa_t *spa, vdev_t *vd, uint64_t prevstate)
 {
-	zfs_post_common(spa, vd, FM_RSRC_CLASS, FM_RESOURCE_STATECHANGE);
+#ifdef _KERNEL
+	nvlist_t *aux;
+
+	/*
+	 * Add optional supplemental keys to payload
+	 */
+	aux = fm_nvlist_create(NULL);
+	if (vd && aux) {
+		char *state;
+
+		switch (vd->vdev_state) {
+		default:
+			state = "UNKNOWN";
+			break;
+		case VDEV_STATE_CLOSED:
+		case VDEV_STATE_OFFLINE:
+			state = "OFFLINE";
+			break;
+		case VDEV_STATE_REMOVED:
+			state = "REMOVED";
+			break;
+		case VDEV_STATE_CANT_OPEN:
+			state = "UNAVAIL";
+			break;
+		case VDEV_STATE_FAULTED:
+			state = "FAULTED";
+			break;
+		case VDEV_STATE_DEGRADED:
+			state = "DEGRADED";
+			break;
+		case VDEV_STATE_HEALTHY:
+			state = "ONLINE";
+			break;
+		}
+		if (vd->vdev_physpath) {
+			(void) nvlist_add_string(aux,
+			    FM_EREPORT_PAYLOAD_ZFS_VDEV_PHYSPATH,
+			    vd->vdev_physpath);
+		}
+		(void) nvlist_add_uint64(aux,
+		    FM_EREPORT_PAYLOAD_ZFS_VDEV_PREVSTATE, prevstate);
+
+		(void) nvlist_add_string(aux,
+		    FM_EREPORT_PAYLOAD_ZFS_VDEV_STATE_STR, state);
+	}
+
+	zfs_post_common(spa, vd, FM_RSRC_CLASS, FM_RESOURCE_STATECHANGE,
```

Comment:
I don't think it would be unreasonable to add all of these keys to events posted with `zfs_post_common()`.  We don't know what end users might want to do and I can't think of a compelling reason not to make this available.

So what if we did this state id to string conversion in user space rather than in the kernel.  We really don't have to worry about these values ever changing.  Not only are some of them stored on-disk in the label, but the libzfs library is built with these values.  We can't change then without breaking user/kernel compatibility.

This could be done one of two ways.
* The ZED could do the translation and add the new keys to the environment using the existing `zpool_state_to_name()` function from libzfs, or
* A `zpool_state_to_name()` shell function could be added to `cmd/zed/zed.d/zed-functions.sh` so any zedlet can easily do the translation.